modules/core/src/main/java/org/apache/synapse/config/xml/SynapsePathFactory.java

```diff
@@ -48,7 +50,22 @@ public static SynapsePath getSynapsePath(OMElement elem, QName attribName)
             if(pathAttrib.getAttributeValue().startsWith("json-eval(")) {
                 path = new SynapseJsonPath(pathAttrib.getAttributeValue().substring(10, pathAttrib.getAttributeValue().length() - 1));
             } else {
-                path = new SynapseXPath(pathAttrib.getAttributeValue());
+                try {
+                    path = new SynapseXPath(pathAttrib.getAttributeValue());
+                } catch (org.jaxen.XPathSyntaxException ex) {
+                    /* Try and see whether the expression can be compiled with XPath 2.0
+                     * This will only be done if the failover DOM XPath 2.0 is enabled */
+                    if (Boolean.TRUE.toString().equalsIgnoreCase(SynapsePropertiesLoader.loadSynapseProperties().
+                            getProperty(SynapseConstants.FAIL_OVER_DOM_XPATH_PROCESSING))) {
+                        if(log.isDebugEnabled()) {
+                            log.debug(
+                                    "Trying to compile the expression in XPath 2.0: " + pathAttrib.getAttributeValue());
+                        }
+                        path = new SynapseXPath(pathAttrib.getAttributeValue(), elem);
+                    } else {
+                        throw ex;
```

Comment:
this can happen even for the invalid XPath expressions. Should we log and throw the exception? And yeah, I will add this to the documents.
Thanks :)